storage/innobase/buf/buf0buf.cc

```diff
@@ -2404,6 +2404,81 @@ buf_zip_decompress(
 	return(FALSE);
 }
 
+ATTRIBUTE_COLD
+/** Try to merge buffered changes to a buffer pool page.
+@param block     buffer-fixed and latched block
+@param rw_latch  RW_X_LATCH, RW_SX_LATCH, RW_S_LATCH held on block
+@param err       error code
+@return whether the page is invalid (corrupted) */
+static bool buf_page_ibuf_merge_try(buf_block_t *block, ulint rw_latch,
+                                    dberr_t *err)
+{
+  ut_ad(block->page.lock.have_any());
+  ut_ad(block->page.buf_fix_count());
+
+  if (fil_page_get_type(block->page.frame) != FIL_PAGE_INDEX ||
+      !page_is_leaf(block->page.frame))
+    return false;
+
+  uint32_t state;
+
+  if (rw_latch != RW_X_LATCH)
+  {
+    if (rw_latch == RW_S_LATCH)
+    {
+      if (!block->page.lock.s_x_upgrade_try())
+      {
+        block->page.lock.s_unlock();
+        block->page.lock.x_lock();
+        state= block->page.state();
+        if (state < buf_page_t::UNFIXED)
+        {
+        fail:
+          block->page.lock.x_unlock();
+          return true;
+        }
+        ut_ad(state & ~buf_page_t::LRU_MASK);
+        ut_ad(state < buf_page_t::READ_FIX);
+        if (state < buf_page_t::IBUF_EXIST || state >= buf_page_t::REINIT)
+          /* ibuf_merge_or_delete_for_page() was already invoked in
+          another thread. */
```

Comment:
Essentially we are re-checking the condition again after acquiring the latch. Looks all good. May be we can bring down the same check from the caller to this function and have some lamda which we can use in the beginning of the function and here.

Reply:
Earlier, you [suggested something similar](https://github.com/MariaDB/server/pull/3022#discussion_r1481081933) and I found out that a lambda function would not be inlined. It might work here, but the emitted code would have to be reviewed on a number of compilers and targets. I would rather keep it ‘ugly’ for now in order to avoid surprises.